File: Data Structures & Algorithms/reorder-linked-list/submission-2.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        ReorderList(head, head.next);
    }

    private ListNode ReorderList(ListNode head, ListNode next) {
        if (next == null) return head;

        ListNode currentHead = ReorderList(head, next.next);
        if (currentHead == null) return null;

        if (currentHead == next || currentHead.next == next) {
            next.next = null;
            return null;
        }

        ListNode temp = currentHead.next;
        currentHead.next = next;
        next.next = temp;
        return temp;
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode head;
        ListNode* temp = &head;
        int left = 0;
        while (l1 != nullptr || l2 != nullptr || left != 0) {
            int num1 = l1 != nullptr ? l1->val : 0;
            int num2 = l2 != nullptr ? l2->val : 0;
            int result = num1 + num2 + left;

            left = result / 10;
            result %= 10;

            temp->next = new ListNode(result);
            temp = temp->next;

            l1 = l1 != nullptr ? l1->next : nullptr;
            l2 = l2 != nullptr ? l2->next : nullptr;
        }

        return head.next;
    }
};
